File: src/main/java/org/moinex/ui/dialog/wallettransaction/BaseTransferManagement.java
/*
 * Filename: AddTransferController.java
 * Created on: October  4, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.wallettransaction;

import java.math.BigDecimal;
import java.util.List;
import java.util.function.Consumer;
import java.util.function.Function;
import javafx.beans.value.ChangeListener;
import javafx.fxml.FXML;
import javafx.scene.control.ComboBox;
import javafx.scene.control.DatePicker;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.Category;
import org.moinex.model.wallettransaction.Transfer;
import org.moinex.model.wallettransaction.Wallet;
import org.moinex.service.CalculatorService;
import org.moinex.service.CategoryService;
import org.moinex.service.I18nService;
import org.moinex.service.WalletService;
import org.moinex.service.WalletTransactionService;
import org.moinex.ui.common.CalculatorController;
import org.moinex.util.Constants;
import org.moinex.util.SuggestionsHandlerHelper;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ConfigurableApplicationContext;

/**
 * Controller for the Add Transfer dialog
 */
@NoArgsConstructor
public abstract class BaseTransferManagement {
    @FXML protected Label senderWalletAfterBalanceValueLabel;

    @FXML protected Label receiverWalletAfterBalanceValueLabel;

    @FXML protected Label senderWalletCurrentBalanceValueLabel;

    @FXML protected Label receiverWalletCurrentBalanceValueLabel;

    @FXML protected ComboBox<Wallet> senderWalletComboBox;

    @FXML protected ComboBox<Wallet> receiverWalletComboBox;

    @FXML protected TextField transferValueField;

    @FXML protected TextField descriptionField;

    @FXML protected DatePicker transferDatePicker;

    @FXML protected ComboBox<Category> categoryComboBox;

    protected ConfigurableApplicationContext springContext;

    protected WalletService walletService;

    protected WalletTransactionService walletTransactionService;

    protected CalculatorService calculatorService;

    protected CategoryService categoryService;

    protected I18nService i18nService;

    protected List<Wallet> wallets;

    protected List<Category> categories;

    private ChangeListener<String> transferValueListener;

    protected SuggestionsHandlerHelper<Transfer> suggestionsHandler;

    @Autowired
    protected BaseTransferManagement(
            WalletService walletService,
            WalletTransactionService walletTransactionService,
            CalculatorService calculatorService,
            CategoryService categoryService,
            I18nService i18nService,
            ConfigurableApplicationContext springContext) {
        this.walletService = walletService;
        this.walletTransactionService = walletTransactionService;
        this.calculatorService = calculatorService;
        this.categoryService = categoryService;
        this.i18nService = i18nService;
        this.springContext = springContext;
    }

    @FXML
    protected void initialize() {
        configureSuggestions();
        configureListeners();
        configureComboBoxes();

        loadWalletsFromDatabase();
        loadSuggestionsFromDatabase();
        loadCategoriesFromDatabase();

        populateComboBoxes();

        // Configure the date picker
        UIUtils.setDatePickerFormat(transferDatePicker, i18nService);

        // Reset all labels
        UIUtils.resetLabel(senderWalletAfterBalanceValueLabel);
        UIUtils.resetLabel(receiverWalletAfterBalanceValueLabel);
        UIUtils.resetLabel(senderWalletCurrentBalanceValueLabel);
        UIUtils.resetLabel(receiverWalletCurrentBalanceValueLabel);

        senderWalletComboBox.setOnAction(
                e -> {
                    updateSenderWalletBalance();
                    updateSenderWalletAfterBalance();
                });

        receiverWalletComboBox.setOnAction(
                e -> {
                    updateReceiverWalletBalance();
                    updateReceiverWalletAfterBalance();
                });
    }

    @FXML
    protected void handleCancel() {
        Stage stage = (Stage) descriptionField.getScene().getWindow();
        stage.close();
    }

    @FXML
    protected abstract void handleSave();

    protected abstract void updateAfterBalance(
            Wallet mainWallet, Wallet otherWallet, Label afterBalanceLabel, boolean isSender);

    @FXML
    protected void handleOpenCalculator() {
        WindowUtils.openPopupWindow(
                Constants.CALCULATOR_FXML,
                i18nService.tr(Constants.TranslationKeys.WALLETTRANSACTION_LABEL_CALCULATOR),
                springContext,
                (CalculatorController controller) -> {},
                List.of(() -> calculatorService.updateComponentWithResult(transferValueField)));
    }

    protected void updateSenderWalletBalance() {
        Wallet senderWt = senderWalletComboBox.getValue();

        if (senderWt == null) {
            return;
        }

        if (senderWt.getBalance().compareTo(BigDecimal.ZERO) < 0) {
            UIUtils.setLabelStyle(
                    senderWalletCurrentBalanceValueLabel, Constants.NEGATIVE_BALANCE_STYLE);
        } else {
            UIUtils.setLabelStyle(
                    senderWalletCurrentBalanceValueLabel, Constants.NEUTRAL_BALANCE_STYLE);
        }

        senderWalletCurrentBalanceValueLabel.setText(UIUtils.formatCurrency(senderWt.getBalance()));
    }

    protected void updateReceiverWalletBalance() {
        Wallet receiverWt = receiverWalletComboBox.getValue();

        if (receiverWt == null) {
            return;
        }

        if (receiverWt.getBalance().compareTo(BigDecimal.ZERO) < 0) {
            UIUtils.setLabelStyle(
                    receiverWalletCurrentBalanceValueLabel, Constants.NEGATIVE_BALANCE_STYLE);
        } else {
            UIUtils.setLabelStyle(
                    receiverWalletCurrentBalanceValueLabel, Constants.NEUTRAL_BALANCE_STYLE);
        }

        receiverWalletCurrentBalanceValueLabel.setText(
                UIUtils.formatCurrency(receiverWt.getBalance()));
    }

    /**
     * Wrapper to update the projected balance of the sender wallet
     */
    protected void updateSenderWalletAfterBalance() {
        updateAfterBalance(
                senderWalletComboBox.getValue(),
                receiverWalletComboBox.getValue(),
                senderWalletAfterBalanceValueLabel,
                true);
    }

    /**
     * Wrapper to update the projected balance of the receiver wallet
     */
    protected void updateReceiverWalletAfterBalance() {
        updateAfterBalance(
                receiverWalletComboBox.getValue(),
                senderWalletComboBox.getValue(),
                receiverWalletAfterBalanceValueLabel,
                false);
    }

    protected void configureListeners() {
        // Update sender wallet after balance when transfer value changes
        transferValueListener =
                (observable, oldValue, newValue) -> {
                    if (!newValue.matches(Constants.MONETARY_VALUE_REGEX)) {
                        transferValueField.setText(oldValue);
                    } else {
                        updateSenderWalletAfterBalance();
                        updateReceiverWalletAfterBalance();
                    }
                };

        transferValueField.textProperty().addListener(transferValueListener);
    }

    protected void loadWalletsFromDatabase() {
        wallets = walletService.getAllNonArchivedWalletsOrderedByName();
    }

    protected void loadCategoriesFromDatabase() {
        categories = categoryService.getNonArchivedCategoriesOrderedByName();
    }

    protected void populateComboBoxes() {
        senderWalletComboBox.getItems().setAll(wallets);
        receiverWalletComboBox.getItems().setAll(wallets);

        // set the first as empty to allow no category
        categories.addFirst(null);

        categoryComboBox.getItems().setAll(categories);
    }

    protected void configureComboBoxes() {
        UIUtils.configureComboBox(senderWalletComboBox, Wallet::getName);
        UIUtils.configureComboBox(receiverWalletComboBox, Wallet::getName);
        UIUtils.configureComboBox(categoryComboBox, Category::getName);
    }

    private void loadSuggestionsFromDatabase() {
        suggestionsHandler.setSuggestions(walletTransactionService.getTransferSuggestions());
    }

    private void configureSuggestions() {
        Function<Transfer, String> filterFunction = Transfer::getDescription;

        // Format:
        //    Description
        //    Amount | From: Wallet | To: Wallet | Category
        Function<Transfer, String> displayFunction =
                tf ->
                        String.format(
                                "%s%n%s | "
                                        + i18nService.tr(
                                                Constants.TranslationKeys
                                                        .WALLETTRANSACTION_SUGGESTION_FROM)
                                        + " %s | "
                                        + i18nService.tr(
                                                Constants.TranslationKeys
                                                        .WALLETTRANSACTION_SUGGESTION_TO)
                                        + " %s | %s ",
                                tf.getDescription(),
                                UIUtils.formatCurrency(tf.getAmount()),
                                tf.getSenderWallet().getName(),
                                tf.getReceiverWallet().getName(),
                                tf.getCategory() != null
                                        ? tf.getCategory().getName()
                                        : i18nService.tr(
                                                Constants.TranslationKeys
                                                        .WALLETTRANSACTION_SUGGESTION_NO_CATEGORY));

        Consumer<Transfer> onSelectCallback = this::fillFieldsWithTransaction;

        suggestionsHandler =
                new SuggestionsHandlerHelper<>(
                        descriptionField, filterFunction, displayFunction, onSelectCallback);

        suggestionsHandler.enable();
    }

    private void fillFieldsWithTransaction(Transfer t) {
        senderWalletComboBox.setValue(t.getSenderWallet());
        receiverWalletComboBox.setValue(t.getReceiverWallet());

        // Deactivate the listener to avoid the event of changing the text of
        // the descriptionField from being triggered. After changing the text,
        // the listener is activated again
        suggestionsHandler.disable();
        descriptionField.setText(t.getDescription());
        suggestionsHandler.enable();

        transferValueField.setText(t.getAmount().toString());

        categoryComboBox.setValue(t.getCategory());

        updateSenderWalletBalance();
        updateSenderWalletAfterBalance();

        updateReceiverWalletBalance();
        updateReceiverWalletAfterBalance();
    }

    public void disableTransferValueListener() {
        if (transferValueListener != null) {
            transferValueField.textProperty().removeListener(transferValueListener);
        }
    }

    public void enableTransferValueListener() {
        if (transferValueListener != null) {
            transferValueField.textProperty().addListener(transferValueListener);
        }
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/wallettransaction/base/BaseTransferManagement.kt
/*
 * Filename: BaseTransferManagement.kt (original filename: AddTransferController.java)
 * Created on: October  4, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 21/03/2026
 */

package org.moinex.ui.dialog.wallettransaction.base

import javafx.beans.value.ChangeListener
import javafx.fxml.FXML
import javafx.scene.control.ComboBox
import javafx.scene.control.DatePicker
import javafx.scene.control.Label
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.moinex.common.constants.Constants
import org.moinex.common.constants.TranslationKeys
import org.moinex.common.helper.SuggestionsHandlerHelper
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.Category
import org.moinex.model.wallettransaction.Transfer
import org.moinex.model.wallettransaction.Wallet
import org.moinex.service.CalculatorService
import org.moinex.service.CategoryService
import org.moinex.service.PreferencesService
import org.moinex.service.wallet.WalletService
import org.moinex.ui.common.CalculatorController
import org.springframework.context.ConfigurableApplicationContext
import java.math.BigDecimal

/**
 * Base class for transfer dialog controllers
 */
abstract class BaseTransferManagement(
    protected val walletService: WalletService,
    protected val calculatorService: CalculatorService,
    protected val categoryService: CategoryService,
    protected val preferencesService: PreferencesService,
    protected val springContext: ConfigurableApplicationContext,
) {
    @FXML
    protected lateinit var senderWalletAfterBalanceValueLabel: Label

    @FXML
    protected lateinit var receiverWalletAfterBalanceValueLabel: Label

    @FXML
    protected lateinit var senderWalletCurrentBalanceValueLabel: Label

    @FXML
    protected lateinit var receiverWalletCurrentBalanceValueLabel: Label

    @FXML
    protected lateinit var senderWalletComboBox: ComboBox<Wallet>

    @FXML
    protected lateinit var receiverWalletComboBox: ComboBox<Wallet>

    @FXML
    protected lateinit var transferValueField: TextField

    @FXML
    protected lateinit var descriptionField: TextField

    @FXML
    protected lateinit var transferDatePicker: DatePicker

    @FXML
    protected lateinit var categoryComboBox: ComboBox<Category?>

    protected lateinit var suggestionsHandler: SuggestionsHandlerHelper<Transfer>

    protected var wallets: List<Wallet> = emptyList()
    protected var categories: List<Category> = emptyList()

    private var transferValueListener: ChangeListener<String>? = null

    @FXML
    protected open fun initialize() {
        configureSuggestions()
        configureListeners()
        configureComboBoxes()

        loadWalletsFromDatabase()
        loadSuggestionsFromDatabase()
        loadCategoriesFromDatabase()

        populateComboBoxes()

        UIUtils.setDatePickerFormat(transferDatePicker)

        UIUtils.resetLabel(senderWalletAfterBalanceValueLabel)
        UIUtils.resetLabel(receiverWalletAfterBalanceValueLabel)
        UIUtils.resetLabel(senderWalletCurrentBalanceValueLabel)
        UIUtils.resetLabel(receiverWalletCurrentBalanceValueLabel)

        senderWalletComboBox.setOnAction {
            updateSenderWalletBalance()
            updateSenderWalletAfterBalance()
        }

        receiverWalletComboBox.setOnAction {
            updateReceiverWalletBalance()
            updateReceiverWalletAfterBalance()
        }
    }

    @FXML
    protected fun handleCancel() {
        val stage = descriptionField.scene.window as Stage
        stage.close()
    }

    @FXML
    protected abstract fun handleSave()

    protected abstract fun updateAfterBalance(
        mainWallet: Wallet?,
        otherWallet: Wallet?,
        afterBalanceLabel: Label,
        isSender: Boolean,
    )

    @FXML
    protected fun handleOpenCalculator() {
        WindowUtils.openPopupWindow(
            Constants.CALCULATOR_FXML,
            preferencesService.translate(TranslationKeys.WALLETTRANSACTION_LABEL_CALCULATOR),
            springContext,
            { _: CalculatorController -> },
            listOf(Runnable { calculatorService.updateComponentWithResult(transferValueField) }),
        )
    }

    protected fun updateSenderWalletBalance() {
        updateWalletBalance(
            senderWalletComboBox.value,
            senderWalletCurrentBalanceValueLabel,
        )
    }

    protected fun updateReceiverWalletBalance() {
        updateWalletBalance(
            receiverWalletComboBox.value,
            receiverWalletCurrentBalanceValueLabel,
        )
    }

    private fun updateWalletBalance(
        wallet: Wallet?,
        balanceLabel: Label,
    ) {
        wallet ?: return

        val style =
            if (wallet.balance < BigDecimal.ZERO) {
                Constants.NEGATIVE_BALANCE_STYLE
            } else {
                Constants.NEUTRAL_BALANCE_STYLE
            }

        UIUtils.setLabelStyle(balanceLabel, style)
        balanceLabel.text = UIUtils.formatCurrency(wallet.balance)
    }

    protected fun updateSenderWalletAfterBalance() {
        updateAfterBalance(
            senderWalletComboBox.value,
            receiverWalletComboBox.value,
            senderWalletAfterBalanceValueLabel,
            true,
        )
    }

    protected fun updateReceiverWalletAfterBalance() {
        updateAfterBalance(
            receiverWalletComboBox.value,
            senderWalletComboBox.value,
            receiverWalletAfterBalanceValueLabel,
            false,
        )
    }

    protected fun configureListeners() {
        transferValueListener =
            ChangeListener { _, oldValue, newValue ->
                if (!newValue.matches(Regex(Constants.MONETARY_VALUE_REGEX))) {
                    transferValueField.text = oldValue
                } else {
                    updateSenderWalletAfterBalance()
                    updateReceiverWalletAfterBalance()
                }
            }

        transferValueField.textProperty().addListener(transferValueListener)
    }

    protected fun loadWalletsFromDatabase() {
        wallets = walletService.getAllNonArchivedWalletsOrderedByName()
    }

    protected fun loadCategoriesFromDatabase() {
        categories = categoryService.getNonArchivedCategoriesOrderedByName().toMutableList()
    }

    protected fun populateComboBoxes() {
        senderWalletComboBox.items.setAll(wallets)
        receiverWalletComboBox.items.setAll(wallets)

        val categoriesWithNull = mutableListOf<Category?>(null)
        categoriesWithNull.addAll(categories)

        categoryComboBox.items.setAll(categoriesWithNull)
    }

    protected fun configureComboBoxes() {
        UIUtils.configureComboBox(senderWalletComboBox) { it.name }
        UIUtils.configureComboBox(receiverWalletComboBox) { it.name }
        UIUtils.configureComboBox(categoryComboBox) { it?.name ?: "" }
    }

    private fun loadSuggestionsFromDatabase() {
        suggestionsHandler.suggestions = walletService.getTransferSuggestions()
    }

    private fun configureSuggestions() {
        val filterFunction: (Transfer) -> String = { it.description ?: "" }

        val displayFunction: (Transfer) -> String = { tf ->
            val categoryName =
                tf.category?.name
                    ?: preferencesService.translate(
                        TranslationKeys.WALLETTRANSACTION_SUGGESTION_NO_CATEGORY,
                    )

            """
            ${tf.description}
            ${UIUtils.formatCurrency(
                tf.amount,
            )} | ${preferencesService.translate(
                TranslationKeys.WALLETTRANSACTION_SUGGESTION_FROM,
            )} ${tf.senderWallet.name} | ${preferencesService.translate(
                TranslationKeys.WALLETTRANSACTION_SUGGESTION_TO,
            )} ${tf.receiverWallet.name} | $categoryName
            """.trimIndent()
        }

        val onSelectCallback: (Transfer) -> Unit = { fillFieldsWithTransaction(it) }

        suggestionsHandler =
            SuggestionsHandlerHelper(
                descriptionField,
                filterFunction,
                displayFunction,
                onSelectCallback,
            )

        suggestionsHandler.enable()
    }

    private fun fillFieldsWithTransaction(t: Transfer) {
        senderWalletComboBox.value = t.senderWallet
        receiverWalletComboBox.value = t.receiverWallet

        suggestionsHandler.disable()
        descriptionField.text = t.description
        suggestionsHandler.enable()

        transferValueField.text = t.amount.toString()
        categoryComboBox.value = t.category

        updateSenderWalletBalance()
        updateSenderWalletAfterBalance()

        updateReceiverWalletBalance()
        updateReceiverWalletAfterBalance()
    }

    fun disableTransferValueListener() {
        transferValueListener?.let {
            transferValueField.textProperty().removeListener(it)
        }
    }

    fun enableTransferValueListener() {
        transferValueListener?.let {
            transferValueField.textProperty().addListener(it)
        }
    }
}
